metrics.go

```diff
@@ -0,0 +1,106 @@
+package aibridge
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var baseLabels []string = []string{"provider", "model"}
+
+const (
+	InterceptionCountStatusFailed    = "failed"
+	InterceptionCountStatusCompleted = "completed"
+)
+
+type Metrics struct {
+	// Interception-related metrics.
+	InterceptionDuration  *prometheus.HistogramVec
+	InterceptionCount     *prometheus.CounterVec
+	InterceptionsInflight *prometheus.GaugeVec
+	PassthroughCount      *prometheus.CounterVec
+
+	// Prompt-related metrics.
+	PromptCount *prometheus.CounterVec
+
+	// Token-related metrics.
+	TokenUseCount *prometheus.CounterVec
+
+	// Tool-related metrics.
+	InjectedToolUseCount    *prometheus.CounterVec
+	NonInjectedToolUseCount *prometheus.CounterVec
+}
+
+// NewMetrics creates AND registers metrics. It will panic if a collector has already been registered.
+// Note: we are not specifying namespace in the metrics; the provided registerer may specify a "namespace"
+// using [prometheus.WrapRegistererWithPrefix].
+func NewMetrics(reg prometheus.Registerer) *Metrics {
+	return &Metrics{
+		// Interception-related metrics.
+
+		// Estimated cardinality: 2 providers, 5 models, 2 statuses, 2 routes, 3 methods = up to 120 PER INITIATOR.
+		InterceptionCount: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "interceptions",
+			Name:      "total",
+			Help:      "The count of intercepted requests.",
+		}, append(baseLabels, "status", "route", "method", "initiator_id")),
+		// Estimated cardinality: 2 providers, 5 models, 2 routes = up to 20.
+		// NOTE: route is not unbounded because this is only for intercepted routes.
+		InterceptionsInflight: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: "interceptions",
+			Name:      "inflight",
+			Help:      "The number of intercepted requests which are being processed.",
+		}, append(baseLabels, "route")),
+		// Estimated cardinality: 2 providers, 5 models, 7 buckets + 3 extra series (count, sum, +Inf) = up to 100.
```

Comment:
Do we expect the cardinality to increase in the future?

Reply:
What do you mean? Add more labels?